File: frontend/de/unika/ipd/grgen/util/VCGDumper.java
/*
 * GrGen: graph rewrite generator tool -- release GrGen.NET 2.6
 * Copyright (C) 2003-2009 Universitaet Karlsruhe, Institut fuer Programmstrukturen und Datenorganisation, LS Goos
 * licensed under LGPL v3 (see LICENSE.txt included in the packaging of this file)
 */

/**
 * @author Sebastian Hack
 * @version $Id$
 */
package de.unika.ipd.grgen.util;

import java.awt.Color;
import java.io.PrintStream;
import java.util.HashMap;

/**
 * A VCG Graph dumper
 */
public class VCGDumper implements GraphDumper {

	/** where to put the graph to */
	private PrintStream ps;

	/** Index in the vcg colormap for user defined colors */
	private int currSetColor;

	/** Prefix for the nodes. */
	private static String prefix = "n";

	private static HashMap<Color, String> colorMap;
	private static HashMap<Integer, String> shapeMap;
	private static HashMap<Integer, String> lineStyleMap;

	static {
		colorMap = new HashMap<Color, String>();
		shapeMap = new HashMap<Integer, String>();
		lineStyleMap = new HashMap<Integer, String>();

		colorMap.put(Color.BLACK, "black");
		colorMap.put(Color.BLUE, "lightblue");
		colorMap.put(Color.CYAN, "cyan");
		colorMap.put(Color.GRAY, "lightgrey");
		colorMap.put(Color.DARK_GRAY, "darkgrey");
		colorMap.put(Color.MAGENTA, "magenta");
		colorMap.put(Color.ORANGE, "orange");
		colorMap.put(Color.GREEN, "green");
		colorMap.put(Color.RED, "red");
		colorMap.put(Color.PINK, "pink");
		colorMap.put(Color.YELLOW, "yellow");
		colorMap.put(Color.WHITE, "white");

		shapeMap.put(new Integer(BOX), "box");
		shapeMap.put(new Integer(RHOMB), "rhomb");
		shapeMap.put(new Integer(ELLIPSE), "ellipse");
		shapeMap.put(new Integer(TRIANGLE), "triangle");

		lineStyleMap.put(new Integer(SOLID), "continuous");
		lineStyleMap.put(new Integer(DASHED), "dashed");
		lineStyleMap.put(new Integer(DOTTED), "dotted");
	}

	/**
	 * Make a string usable for output.
	 * This escapes anything that has to be escaped.
	 * @param s The input string.
	 * @return A string ready for dumping.
	 */
	private static String escapeString(String s) {
		return s.replaceAll("\"", "\\\\\"");
	}

	/**
	 * Make a new VCG dumper.
	 * @param ps The print stream to dump the graph to.
	 */
	public VCGDumper(PrintStream ps) {
		this.ps = ps;
		this.currSetColor = 32;
	}

	/**
	 * Dump graph preamble.
	 */
	public void begin() {
		ps.println("graph:{\nlate_edge_labels:yes\ndisplay_edge_labels:yes\n"
					   + "manhattan_edges:yes\nport_sharing:no\n");
	}

	/**
	 * Dump epilog.
	 * @see de.unika.ipd.grgen.util.GraphDumper#finish()
	 */
	public void finish() {
		ps.println("}");
		ps.flush();
		ps.close();
	}

	/**
	 * Get a VCG color for a Java color.
	 * @param col The Java color.
	 * @return The VCG color.
	 */
	private String getColor(Color col) {
		String res;

		if(colorMap.containsKey(col))
			res = colorMap.get(col);
		else if(currSetColor < 256) {
			// Get the current index and increment it
			int index = currSetColor++;

			// Convert it to a string and put in the color map
			res = String.valueOf(index);
			colorMap.put(col, res);

			// issue a vcg colormap statement
			ps.println("colorentry " + index + ": " +
						   col.getRed() + " " + col.getGreen() + " " + col.getBlue());
		}
		else
			res = "white";

		return res;
	}

	private String getPrefix() {
		return prefix;
	}

	/**
	 * Make a VCG string from the node's attributes.
	 * @param d The node to dump.
	 * @return VCG statements describing the node.
	 */
	private String getNodeAttributes(GraphDumpable d) {
		String col = getColor(d.getNodeColor());
		Integer shp = new Integer(d.getNodeShape());

		String info = d.getNodeInfo();
		if(info != null)
			info = escapeString(info);

		String label = escapeString(d.getNodeLabel());

		String s = "title:\"" + getPrefix() + d.getNodeId()
			+ "\" label:\"" + label + "\"";

		if(info != null)
			s += " info1:\"" + info + "\"";
		s += " color:" + col;
		if(shapeMap.containsKey(shp))
			s += " shape:" + shapeMap.get(shp);

		return s;
	}

	public void node(GraphDumpable d) {
		ps.println("node:{" + getNodeAttributes(d) + "}");
	}

	public void edge(GraphDumpable from, GraphDumpable to, String label,
					 int style, Color color) {
		if(from!=null && to !=null) {
			String col = getColor(color);

			String s = "edge:{sourcename:\"" + getPrefix() + from.getNodeId()
				+ "\" targetname:\"" + getPrefix() + to.getNodeId() + "\"";

			if(label != null)
				s += " label:\"" + escapeString(label) + "\"";

			s += " color:" + col;

			if(style != DEFAULT)
				s += " linestyle:" + lineStyleMap.get(new Integer(style));

			s += "}";

			ps.println(s);
		}
	}

	public void edge(GraphDumpable from, GraphDumpable to, String label, int style) {
		edge(from, to, label, style, Color.BLACK);
	}

	public void edge(GraphDumpable from, GraphDumpable to, String label) {
		edge(from, to, label, DEFAULT, Color.BLACK);
	}

	public void edge(GraphDumpable from, GraphDumpable to) {
		edge(from, to, null, DEFAULT, Color.BLACK);
	}

	/**
	 * @see de.unika.ipd.grgen.util.GraphDumper#beginSubgraph(java.lang.String)
	 */
	public void beginSubgraph(GraphDumpable d) {
		ps.println("graph:{" + getNodeAttributes(d)
					   + " status:clustered");
	}

	public void beginSubgraph(String title) {
		ps.print("graph:{title:\"");
		ps.print(title);
		ps.println('\"');
		ps.print("  label:\"");
		ps.print(title);
		ps.println('\"');
		ps.println("  status:clustered");
	}

	/**
	 * @see de.unika.ipd.grgen.util.GraphDumper#endSubgraph()
	 */
	public void endSubgraph() {
		ps.println("}\n");
	}

}

File: frontend/de/unika/ipd/grgen/util/VCGDumper.cs
using System.Collections.Generic;

/*
 * GrGen: graph rewrite generator tool -- release GrGen.NET 8.1
 * Copyright (C) 2003-2026 Universitaet Karlsruhe, Institut fuer Programmstrukturen und Datenorganisation, LS Goos; and free programmers
 * licensed under LGPL v3, some components/parts use different licenses (see LICENSE.txt included in the packaging of this file)
 * www.grgen.de / www.grgen.net
 */

/// <summary>
/// @author Sebastian Hack
/// </summary>

namespace de.unika.ipd.grgen.util
{

	/// <summary>
	/// A VCG Graph dumper
	/// </summary>
	public class VCGDumper : GraphDumper
	{
		/// <summary>
		/// where to put the graph to </summary>
		private PrintStream ps;

		/// <summary>
		/// Index in the vcg colormap for user defined colors </summary>
		private int currSetColor;

		/// <summary>
		/// Prefix for the nodes. </summary>
		private static string prefix = "n";

		private static Dictionary<Color, string> colorMap;
		private static Dictionary<Shape, string> shapeMap;
		private static Dictionary<Style, string> lineStyleMap;

		static VCGDumper()
		{
			colorMap = new Dictionary<Color, string>();
			shapeMap = new Dictionary<Shape, string>();
			lineStyleMap = new Dictionary<Style, string>();

			colorMap[Color.BLACK] = "black";
			colorMap[Color.BLUE] = "lightblue";
			colorMap[Color.CYAN] = "cyan";
			colorMap[Color.GRAY] = "lightgrey";
			colorMap[Color.DARK_GRAY] = "darkgrey";
			colorMap[Color.MAGENTA] = "magenta";
			colorMap[Color.ORANGE] = "orange";
			colorMap[Color.GREEN] = "green";
			colorMap[Color.RED] = "red";
			colorMap[Color.PINK] = "pink";
			colorMap[Color.YELLOW] = "yellow";
			colorMap[Color.WHITE] = "white";

			shapeMap[Shape.BOX] = "box";
			shapeMap[Shape.RHOMB] = "rhomb";
			shapeMap[Shape.ELLIPSE] = "ellipse";
			shapeMap[Shape.TRIANGLE] = "triangle";

			lineStyleMap[Style.SOLID] = "continuous";
			lineStyleMap[Style.DASHED] = "dashed";
			lineStyleMap[Style.DOTTED] = "dotted";
		}

		/// <summary>
		/// Make a string usable for output.
		/// This escapes anything that has to be escaped. </summary>
		/// <param name="s"> The input string. </param>
		/// <returns> A string ready for dumping. </returns>
		private static string EscapeString(string s)
		{
			return s.ReplaceAll("\"", "\\\\\"");
		}

		/// <summary>
		/// Make a new VCG dumper. </summary>
		/// <param name="ps"> The print stream to dump the graph to. </param>
		public VCGDumper(PrintStream ps)
		{
			this.ps = ps;
			this.currSetColor = 32;
		}

		/// <summary>
		/// Dump graph preamble.
		/// </summary>
		public virtual void Begin()
		{
			ps.Println("graph:{\nlate_edge_labels:yes\ndisplay_edge_labels:yes\n"
					+ "manhattan_edges:yes\nport_sharing:no\n");
		}

		/// <summary>
		/// Dump epilog. </summary>
		/// <seealso cref="de.unika.ipd.grgen.util.GraphDumper.finish()"/>
		public virtual void Finish()
		{
			ps.Println("}");
			ps.Flush();
			ps.Close();
		}

		/// <summary>
		/// Get a VCG color for a Java color. </summary>
		/// <param name="col"> The Java color. </param>
		/// <returns> The VCG color. </returns>
		private string GetColor(Color col)
		{
			string res;

			if(colorMap.ContainsKey(col))
				res = colorMap[col];
			else if(currSetColor < 256)
			{
				/*
				// Get the current index and increment it
				int index = currSetColor++;

				// Convert it to a string and put in the color map
				res = index.ToString();
				colorMap[col] = res;

				// issue a vcg colormap statement
				ps.Println("colorentry " + index + ": " +
						col.GetRed() + " " + col.GetGreen() + " " + col.GetBlue());
				*/
				throw new System.NotImplementedException(); // removed during the porting to C#, re-implement when needed (with System.Drawing.Color) - but using de.unika.ipd.grGen.libGr.GrColor should be sufficient
			}
			else
				res = "white";

			return res;
		}

		private static string Prefix
		{
			get
			{
				return prefix;
			}
		}

		/// <summary>
		/// Make a VCG string from the node's attributes. </summary>
		/// <param name="d"> The node to dump. </param>
		/// <returns> VCG statements describing the node. </returns>
		private string GetNodeAttributes(GraphDumpable d)
		{
			string col = GetColor(d.NodeColor);
			Shape shp = d.NodeShape;

			string info = d.NodeInfo;
			if(!string.ReferenceEquals(info, null))
				info = EscapeString(info);

			string label = EscapeString(d.NodeLabel);

			string s = "title:\"" + Prefix + d.NodeId
					+ "\" label:\"" + label + "\"";

			if(!string.ReferenceEquals(info, null))
				s += " info1:\"" + info + "\"";
			s += " color:" + col;
			if(shapeMap.ContainsKey(shp))
				s += " shape:" + shapeMap[shp];

			return s;
		}

		public virtual void Node(GraphDumpable d)
		{
			ps.Println("node:{" + GetNodeAttributes(d) + "}");
		}

		public virtual void Edge(GraphDumpable from, GraphDumpable to, string label,
				Style style, Color color)
		{
			if(from != null && to != null)
			{
				string col = GetColor(color);

				string s = "edge:{sourcename:\"" + Prefix + from.NodeId
						+ "\" targetname:\"" + Prefix + to.NodeId + "\"";

				if(!string.ReferenceEquals(label, null))
					s += " label:\"" + EscapeString(label) + "\"";

				s += " color:" + col;

				if(style != Style.DEFAULT)
					s += " linestyle:" + lineStyleMap[style];

				s += "}";

				ps.Println(s);
			}
		}

		public virtual void Edge(GraphDumpable from, GraphDumpable to, string label, Style style)
		{
			Edge(from, to, label, style, Color.BLACK);
		}

		public virtual void Edge(GraphDumpable from, GraphDumpable to, string label)
		{
			Edge(from, to, label, Style.DEFAULT, Color.BLACK);
		}

		public virtual void Edge(GraphDumpable from, GraphDumpable to)
		{
			Edge(from, to, null, Style.DEFAULT, Color.BLACK);
		}

		/// <seealso cref="de.unika.ipd.grgen.util.GraphDumper.beginSubgraph(java.lang.String)"/>
		public virtual void BeginSubgraph(GraphDumpable d)
		{
			ps.Println("graph:{" + GetNodeAttributes(d)
					+ " status:clustered");
		}

		public virtual void BeginSubgraph(string title)
		{
			ps.Print("graph:{title:\"");
			ps.Print(title);
			ps.Println('\"');
			ps.Print("  label:\"");
			ps.Print(title);
			ps.Println('\"');
			ps.Println("  status:clustered");
		}

		/// <seealso cref="de.unika.ipd.grgen.util.GraphDumper.endSubgraph()"/>
		public virtual void EndSubgraph()
		{
			ps.Println("}\n");
		}
	}

}
